Implement GetTypesByMetadataName

333fred opened the pull request:
Closes https://github.com/dotnet/roslyn/issues/57802. Implements the GetTypesByMetadataName API, which returns all named types in the current compilation and referenced assemblies that match the given CLR metatdata name. Also updates the IDE's GetBestTypeByMetadataName to use this API under the hood, rather than the existing manual walk they were doing, simplyfing that codepath and proving out the API itself.

Youssef1313 commented on src/Compilers/Core/Portable/Compilation/Compilation.cs:
```diff
@@ -1038,15 +1041,58 @@ public INamedTypeSymbol CreateNativeIntegerTypeSymbol(bool signed)
-#pragma warning disable RS0026 // Do not add multiple public overloads with optional parameters
+        /// <summary>
+        /// Gets all types with the compilation's assembly and all referenced assemblies that have the
+        /// given canonical CLR metadta name.
```
```suggestion
        /// given canonical CLR metadata name.
```

AlekseyTs commented on src/Compilers/Core/Portable/Compilation/Compilation.cs:
```diff
@@ -1025,28 +1025,101 @@ public INamedTypeSymbol CreateNativeIntegerTypeSymbol(bool signed)
+        /// This lookup follows the following order:
+        /// <list type="number">
+        /// <item><description>If the type is found in the compilation's assembly, that type is returned.</description></item>
+        /// <item><description>Next, the core library (the library that defines <c>System.Object</c>) is searched. If the type is found there, that type is returned.</description></item>
```
>the library that defines <c>System.Object</c> [](http://example.com/codeflow?start=55&length=45)

```the library that defines <c>System.Object</c> and has no assembly references```?

AlekseyTs commented on src/Compilers/Core/Portable/Compilation/Compilation.cs:
```diff
@@ -1025,28 +1025,101 @@ public INamedTypeSymbol CreateNativeIntegerTypeSymbol(bool signed)
+        /// Because accessibility to the current assembly is ignored when searching for types that match the provided metadata name, if multiple referenced
+        /// assemblies define the same type symbol (as often happens when users copy well-known types from the BCL or other sources) then this API will return null,
+        /// even if all but one of those symbols would be otherwise inaccessible to user-written code in the current assembly. If it is not an error for a type to exist
+        /// in multiple referenced assemblies, consider using <see cref="GetTypesByMetadataName(string)" /> instead and filtering the results for the symbol required.
```
> If it is not an error for a type to exist in multiple referenced assemblies

Consider if it would be better to re-phrase this part like: "For fine grain control over ambiguity resolution```

AlekseyTs commented on src/Compilers/Core/Portable/Compilation/Compilation.cs:
```diff
@@ -1025,28 +1025,101 @@ public INamedTypeSymbol CreateNativeIntegerTypeSymbol(bool signed)
+        /// This lookup follows the following order:
+        /// <list type="number">
+        /// <item><description>If the type is found in the compilation's assembly, that type is returned.</description></item>
+        /// <item><description>Next, the core library (the library that defines <c>System.Object</c>) is searched. If the type is found there, that type is returned.</description></item>
```
>the core library [](http://example.com/codeflow?start=37&length=16)

It would be good to clarify if the core library would be searched if it is references via an extern alias.

333fred replied:
Clarified this in the remarks below.

AlekseyTs commented on src/Compilers/Core/Portable/Compilation/Compilation.cs:
```diff
@@ -1025,28 +1025,101 @@ public INamedTypeSymbol CreateNativeIntegerTypeSymbol(bool signed)
+        /// <returns>Null if the type can't be found or there was an ambiguity during lookup.</returns>
         /// <remarks>
         /// Since VB does not have the concept of extern aliases, it considers all referenced assemblies.
+        /// <br/>
```
><br/> [](http://example.com/codeflow?start=12&length=5)

It might be worth mentioning that within each assembly the search is performed in each module based on their position in the modules list until a match is found, remaining modules are not examined.

Also, I think type forwarders are ignored.

AlekseyTs commented on src/Compilers/Core/Portable/Compilation/Compilation.cs:
```diff
@@ -1025,28 +1025,101 @@ public INamedTypeSymbol CreateNativeIntegerTypeSymbol(bool signed)
+        /// Gets all types with the compilation's assembly and all referenced assemblies that have the
+        /// given canonical CLR metadata name. Accessibility to the current assembly is ignored when
+        /// searching for matching type names.
+        /// </summary>
```
></summary> [](http://example.com/codeflow?start=12&length=10)

Similar comment about how modules are searched is applicable, I think.

AlekseyTs commented on src/Compilers/Core/Portable/Compilation/Compilation.cs:
```diff
@@ -1025,28 +1025,101 @@ public INamedTypeSymbol CreateNativeIntegerTypeSymbol(bool signed)
+                var result = _getTypesCache.TryAdd(fullyQualifiedMetadataName, val);
+                Debug.Assert(result
+                    || (_getTypesCache.TryGetValue(fullyQualifiedMetadataName, out var addedArray)
+                        && addedArray.Zip(val, (added, calculated) => (added, calculated)).All(el => ReferenceEquals(el.added, el.calculated))));
```
>Zip [](http://example.com/codeflow?start=38&length=3)

What does ```Zip``` do when there is a length mismatch between collections?

RikkiGibson replied:
> The method merges each element of the first sequence with an element that has the same index in the second sequence. If the sequences do not have the same number of elements, the method merges sequences until it reaches the end of one of them. For example, if one sequence has three elements and the other one has four, the result sequence will have only three elements.

https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.zip?view=net-6.0#system-linq-enumerable-zip-3(system-collections-generic-ienumerable((-0))-system-collections-generic-ienumerable((-1))-system-func((-0-1-2)))

It could be simpler to use `Enumerable.SequenceEqual` and pass `ReferenceEqualityComparer.Instance`.

AlekseyTs replied:
It sounds like the assert should be strengthened to check for the same length of both arrays.

333fred replied:
Switched to `SequenceEqual`

AlekseyTs commented on src/Compilers/Core/Portable/Compilation/Compilation.cs:
```diff
@@ -1025,28 +1025,101 @@ public INamedTypeSymbol CreateNativeIntegerTypeSymbol(bool signed)
+                addIfNotNull(Assembly.GetTypeByMetadataName(fullyQualifiedMetadataName));
+
+                var corLib = (IAssemblySymbol)((IAssemblySymbolInternal)Assembly).CorLibrary;
+
```
I am not sure if the order is such a big deal that it is worth the trouble of dealing with core library separately. In any case, I think it would be good avoiding a usage of internal APIs in implementation of this method. I think we could simply use ```ObjectType.ContainingAssembly``` instead.

AlekseyTs commented on src/Compilers/Core/Portable/Compilation/Compilation.cs:
```diff
@@ -1025,28 +1025,101 @@ public INamedTypeSymbol CreateNativeIntegerTypeSymbol(bool signed)
+
+                addIfNotNull(Assembly.GetTypeByMetadataName(fullyQualifiedMetadataName));
+
+                var corLib = (IAssemblySymbol)((IAssemblySymbolInternal)Assembly).CorLibrary;
```
>(IAssemblySymbolInternal)Assembly [](http://example.com/codeflow?start=47&length=33)

It looks like the cast is illegal

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Compilation/CompilationAPITests.cs:
```diff
@@ -3036,5 +3036,213 @@ void M1()
+            AssertEx.Equal("N.C<T>", types[0].ToTestDisplayString());
+            Assert.Same(comp.Assembly.GetPublicSymbol(), types[0].ContainingAssembly);
+            AssertEx.Equal("N.C<T>", types[1].ToTestDisplayString());
+            if (useMetadataReference)
```
>useMetadataReference [](http://example.com/codeflow?start=16&length=20)

I think we can avoid conditional logic around the asserts. There is an API to get an assembly symbol for a given metadata reference instance. That is the one that is supposed to contain the second type regardless of the reference kind. This is probably applicable to other tests as well.

AlekseyTs commented on src/Compilers/VisualBasic/Test/Symbol/CompilationAPITests.vb:
```diff
@@ -0,0 +1,176 @@
+
+Namespace Microsoft.CodeAnalysis.VisualBasic.UnitTests
+
+    Public Class CompilationAPITests
```
>CompilationAPITests [](http://example.com/codeflow?start=17&length=19)

I am assuming that similar comments as for C# tests are applicable to VB tests.

AlekseyTs commented:
Done with review pass (commit 5).

333fred commented:
@AlekseyTs addressed your feedback.

---
In reply to: [1007034424](https://github.com/dotnet/roslyn/pull/58366#issuecomment-1007034424) [](http://example.com/codeflow?ancestors=1007034424,1007034424)

RikkiGibson commented on src/Compilers/VisualBasic/Test/Symbol/CompilationAPITests.vb:
```diff
@@ -0,0 +1,226 @@
+        Inherits BasicTestBase
+
+        <Fact()>
+        Public Sub GetTypesByMetadtaName_NotInSourceNotInReferences()
```
nit: spelling, here and below

AlekseyTs reviewed (approved):
LGTM (commit 8)
